### DISCUSSION TITLE
Reference to Plugin service from an Analyzer

### DISCUSSION THREAD

**Discussion by cyberkaida** [NationalSecurityAgency/ghidra#6582] [Q&A] [answered]
Hi! I have a few questions!

I am working on a plugin that provides some UI and an analyser. My plugin subclasses `ProgramPlugin` and my analyzer subclasses `Analyzer` (also tried with `AbstractAnalyzer`). The Analyzer appears correctly in the analysis and the plugin works, but I would like to be able to expose a service from my plugin to the analyzer.

I can't find a way to get from my Analyzer instance to the plugin's service. I don't have a reference to a `PluginTool` or anything from the analyzer, I don't think I can create and register the analyzer myself (the class finder does this), so I can't pass a reference to my service.

I could use a singleton pattern, but my plugin is tied to the lifetime of the program I think, so I want to avoid that. I think I am not understanding the lifetime of an Analyzer correctly...

If someone has some advice I would really appreciate it! I spent some time looking at the APIs and how analyzers and plugins are registered, but I think I am missing something.

I searched and I only found people talking about their analyzer not appearing because the loader did not find it (typo in the class name).

---

```java
public class RevaFunctionAnalyzer implements Analyzer {

    @Override
    public boolean added(Program program, AddressSetView changed, TaskMonitor monitor, MessageLog log)
            throws CancelledException {
        program.getFunctionManager().getFunctions(changed, true).forEachRemaining((function) -> {
            // TODO: get a reference to the service my Plugin advertises here
        });
        return true;
    }
```

**Comment by astrelsky** [comment]
This probably isn't the recommended way to go about it, but you can give this a shot.
```java
public static PluginTool getTool(DomainObject obj) {
	for (Object o : obj.getConsumerList()) {
		if (o instanceof PluginTool) {
			return (PluginTool) o;
		}
	}
	return null;
}
```

Use that function with a `Program` and then use `Tool::getService` if the result isn't null.

**Comment by dragonmacher** [comment, marked as answer]
It is a shame that we did not supply a `ServiceProvider` to the analzyers.  But, you can do this:

```
		AutoAnalysisManager analysisManager = AutoAnalysisManager.getAnalysisManager(program);
		PluginTool tool = analysisManager.getAnalysisTool();
```

**Reply by cyberkaida** [reply]
Thank you both! I try these suggestions!

**Reply by cyberkaida** [reply]
This worked great! Thank you!
